File: Gombang/[LeetCode75] - [DFS]/1448_Count_Good_Nodes_in_Binary_Tree.cpp
class Solution {
public:

    int value = 0;

    void dfs(TreeNode* root, int maximum)
    {
        if (root == nullptr)
            return;

        if (root->val >= maximum)
        {
            value++;
            maximum = root->val;
        }

        dfs(root->left, maximum);
        dfs(root->right, maximum);
    }

    int goodNodes(TreeNode* root) {
        // 서브트리의 root노드까지 최대값을 계속 비교해준다.

        dfs(root, -10000);

        return value;
    }
};

File: Seori/LeetCode 75/Binary Tree_DFS/1448. Count Good Nodes in Binary Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def goodNodes(self, root: TreeNode) -> int:
        
        def dfs(root: TreeNode, value: int) -> int:
            # 재귀 종료조건
            if root == None:
                return 0

            # 최대값 value를 체크하면서 count를 갱신하고
            count = 0
            if root.val >= value:
                value = root.val
                count += 1
            
            # 이진트리 전위순회
            count += dfs(root.left, value)
            count += dfs(root.right, value)
        
            return count
            
        return dfs(root, root.val)